### PR TITLE
Set infotext for dropped item stacks.

### PR DESCRIPTION
**Pull request by Rogier-5**
This allows the player to identify the item even if several nodes have the same texture, and to see how many items are in the stack.

### CODE REVIEW

**Comment by nerzhul** [comment]
:+1:

**Comment by xunto** [comment]
I don't like it. I understand that some games may need to show infotext but it definitely not a common behavior.

**Comment by Rogier-5** [comment]
Rebased to latest minetest tree

@xunto: it's primarily for convenience. It allows the player to know exactly what type of node is lying on the ground, and how many of it are in the stack.

Alternatively, it could be made configurable, so that server operators can turn it off if they want.

**Comment by Ferk** [comment]
itemstring is not something to display to players.
The "name" is more of an id for internals use, not intended to be shown to the user, moreover the itemstring also contains the metadata of the object. Try dropping a written book.

If there's to be an infotext for the dropped items it should be using the "description" field of the object instead.

File: builtin/game/item_entity.lua [hunk 1]
```diff
@@ -121,6 +123,7 @@ core.register_entity(":__builtin:item", {
 			local name = stack:get_name()
 			if not overflow then
 				obj.itemstring = name .. " " .. count
+				obj.object:set_properties({infotext = obj.itemstring})
```
Comment (rubenwardy) [inline builtin/game/item_entity.lua]: Double quotes please

File: builtin/game/item_entity.lua [hunk 2]
```diff
@@ -112,6 +113,7 @@ core.register_entity(":__builtin:item", {
 				count = count - max_count
 			else
 				self.itemstring = ''
+				self.object:set_properties({infotext = ''})
```
Comment (rubenwardy) [inline builtin/game/item_entity.lua]: Double quotes please

**Comment by Rogier-5** [comment]
I think @ferk is right: this is a player-facing feature, so the description is better to use. Having the itemstack string though, has its uses too.

I made the changes requested by @rubenwardy, and I made the infotext configurable, so that the server operator has a choice:
- use the item description (e.g. 'Dirt (45)')
- use the itemstack text (e.g. 'default:dirt 22')
- use them both
- disable altogether.

**Comment by Ferk** [comment]
Is it worth it for it to be an option? the fact that the itemstring includes the whole metadata reduces the usefulness of that option to just some test environment.

Why not have `make_infotext` be a method inside the entity def table for `__builtin:item`? this way if someone really wants to make the infotext be the itemstring (or any other string) he can just customize it however he likes by overriding `minetest.registered_entities["__builtin:item"].make_infotext(self)` in his testing devtools mod.

**Comment by Rogier-5** [comment]
> the fact that the itemstring includes the whole metadata reduces the usefulness of that option to just some test environment.

The intent was to include just the entity name (e.g. default:dirt) and the count. Maybe I could add an indicator that there is metadata, but the full metadata would indeed be a bit too much in general...

> Why not have make_infotext be a method inside the entity def table for __builtin:item?

That is a good idea ! But I still think that some basic alternatives, (the four I implemented would cover the needs of probably over 99% of players), should be built in. Not everybody is prepared or able to write a mod to improve the text, and if they are, they'd first have to know of the possibility. And as there no (?) standard places where this could be documented as far as I am aware, people will be unlikely to find out by themselves.

**Comment by Ferk** [comment]
> The intent was to include just the entity name (e.g. default:dirt) and the count. Maybe I could add an indicator that there is metadata, but the full metadata would indeed be a bit too much in general...

The itemstring includes the full metadata already, currently. You can see this by dropping a written book while using the stack option.
And I agree that it's too much. What i was saying is that having the metadata included makes that option only suitable for testing anyway, which is why i was thinking it might not be not worth it for it to be optional.

File: builtin/game/item_entity.lua [hunk 3]
```diff
@@ -11,6 +11,54 @@ function core.spawn_item(pos, item)
 	return obj
 end
 
+local item_entity_infotext_error_reported = ""
+local function make_infotext(item, count)
```
Comment (rubenwardy) [inline builtin/game/item_entity.lua]: this is way too complicated all it needs to be is this:

```lua
local stack_infotext_setting = minetest.settings:get_int("item_entity_infotext")
local function make_infotext(stack)
    if stack_infotext_setting == 1 then
        local def = stack:get_definition()
        return def and def.description or stack:get_name()
    elseif stack_infotext_setting == 2 then
        return stack:to_string()
    else
        return ""
    end
end
```

**Comment by SmallJoker** [comment]
Can be done using a mod: (untested)
```Lua
local entity_def = minetest.registered_entities["__builtin:item"]
local old_set_item = entity_def.set_item

function entity_def.set_item(self, itemstring)
	old_set_item(self, itemstring)
	self.object:set_properties({
		-- Use self.itemstring as the function argument
		-- could be outdated due old_set_item
		infotext = make_infotext(self.itemstring)
	})
end
```
Please reopen this PR once it's rebased, commits addressed and still wanted by players to be shipped in builtin.

**Pull request closed without merging**